### PR TITLE
HDX datasets

### PR DESCRIPTION
**Pull request by attiks**
Refs: #HDX-8179

### CODE REVIEW

File: html/modules/custom/hr_paragraphs/src/Controller/HdxController.php [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    $body = $response->getBody() . '';
+    $results = json_decode($body, TRUE);
+
+    return $results['result'];
+  }
+
+  /**
+   * Build Hdx objects.
+   */
+  public function buildHdxObjects(array $results) : array {
```
Comment (attiks) [inline html/modules/custom/hr_paragraphs/src/Controller/HdxController.php:264]: @rupl list of available fields in twig

File: html/modules/custom/hr_paragraphs/hr_paragraphs.module [hunk 2]
```diff
@@ -577,7 +570,9 @@ function hr_paragraphs_preprocess_paragraph__upcoming_events(&$variables) {
   $events = $ical_controller->getIcalEvents($group, $start, $end);
 
   // Sort by start time.
-  usort($events, "__start_sort");
+  usort($events, function ($a, $b) {
+    return strtotime($b['start']) < strtotime($a['start']) ? 1 : -1;
+  });
```
Comment (rupl) [inline html/modules/custom/hr_paragraphs/hr_paragraphs.module:575]: :bowtie:

File: html/modules/custom/hr_paragraphs/src/Controller/HdxController.php (hunk 1 above)
Comment (rupl) [reply html/modules/custom/hr_paragraphs/src/Controller/HdxController.php:264]: Nice it looks like I can probably cherry-pick the templates on top of this. We picked the same twig variable names in our two branches.

File: html/modules/custom/hr_paragraphs/src/Controller/ParagraphController.php [hunk 3]
```diff
@@ -320,64 +327,67 @@ public function getPages($group) {
...
-    if ($request->query->has('page')) {
-      $offset = $request->query->getInt('page', 0) * $limit;
+      // Redirect external links.
+      if ($link->isExternal()) {
+        return new TrustedRedirectResponse($link->getUrl()->getUri());
+      }
     }
 
-    if ($group->field_hdx_country->isEmpty()) {
+    if ($group->field_hdx_dataset_link->isEmpty()) {
```
Comment (rupl) [inline html/modules/custom/hr_paragraphs/src/Controller/ParagraphController.php:340]: When I pulled develop and integrated all this, I am seeing a WSOD. When I looked at logs I got this error:

```
PHP message: Error: Call to a member function isEmpty() on null in /var/www/html/modules/custom/hr_paragraphs/src/Controller/ParagraphController.php on line 340
```

I see the block above and tried adding to the conditional:

```php
    if (!$group->hasField('field_hdx_dataset_link') || $group->field_hdx_dataset_link->isEmpty()) {
      return [
        '#type' => 'markup',
        '#markup' => $this->t('HDX country not chosen.'),
      ];
    }
```

But now I'm seeing the "HDX country not chosen" text, even though in the backend I have Afghanistan chosen. I tried removing it and re-saving the Operation.

**Pull request merged**